src/aggregator/aggregation/gauge.go

```diff
@@ -64,6 +64,17 @@ func (g *Gauge) Update(timestamp time.Time, value float64, annotation []byte) {
 
 	g.count++
 
+	// Keep the last annotation which was set.
+	if len(annotation) > 0 {
+		if cap(g.annotation) < len(annotation) {
+			// Twice as long in case another one comes in
+			// and we could avoid realloc as long as less than this first alloc.
+			g.annotation = make([]byte, 0, 2*len(annotation))
+		}
+		// Reuse any previous allocation while taking a copy.
+		g.annotation = append(g.annotation[:0], annotation...)
```

Comment:
Can we be sure reusing of slice is safe here? I.e. isn't there a case when this reusing will invalidate the slice which has been passed to a protobuf struct (by reference), but not marshalled yet?

Reply:
While I don't have the whole picture and it might be useful to loop in someone else, but after a quick look-around it seems that writes and reads/flushes are mutually exclusive. E.g. `X ∈ {Counter, Gauge, Timer}` is being wrapped in `lockedXAggregation`:

https://github.com/m3db/m3/blob/584f5d4bd03f82153a932a5e9288546c22343598/src/aggregator/aggregator/gauge_elem_gen.go#L43-L49

which is locked during updates

https://github.com/m3db/m3/blob/584f5d4bd03f82153a932a5e9288546c22343598/src/aggregator/aggregator/gauge_elem_gen.go#L151-L157

and during flushes

https://github.com/m3db/m3/blob/584f5d4bd03f82153a932a5e9288546c22343598/src/aggregator/aggregator/gauge_elem_gen.go#L251-L273

Also, the things inside seem to be handled in synchronous way and the `Counter` / `Gauge` / `Timer` doesn't seem to be reused afterwards.